### ISSUE TITLE
Tooltip for function overloaded to mimic variadic types infers wrong overload when called with trailing comma

### ISSUE THREAD

**Issue by krryan** [microsoft/TypeScript#23738]
**TypeScript Version:**  2.7.0-dev.20180418

**Search Terms:** Intellisense overload, tooltip overload

**Code**

```ts
function test<A>(a: A): true;
function test<A, B>(a: A, b: B): false;
function test<T>(...values: T[]): boolean {
    return values.length === 1;
}

declare function assertTrue(truth: true): void;

const noComma = test( // test<string>(a: string): true
    'foo'
)
assertTrue(noComma);

const withComma = test( // test<string, {}>(a: string, b: {}): false, but withComma: true
    'foo',
);
assertTrue(withComma);

assertTrue(test( // test<string, {}>(a: string, b: {}): false, but no error
    'foo',
));
```

**Expected behavior:**
The inferred version of `test` (as shown in hover-over tooltip) for `noComma`, `withComma`, and directly passed to `assertTrue` is the `test<A>(a: A): true` version.

**Actual behavior:**
The inferred version of `test` for the `withComma` and directly-passed versions is the `test<A, B>(a: A, b: B): false`, specifically `test<string, {}>(a: string, b: {}): false`.

Despite this, `withComma` has type `true`, and no error is reported for the call to `assertTrue` with it or when directly passed.

**Playground Link:** [here](http://www.typescriptlang.org/play/index.html#src=function%20test%3CA%3E(a%3A%20A)%3A%20true%3B%0D%0Afunction%20test%3CA%2C%20B%3E(a%3A%20A%2C%20b%3A%20B)%3A%20false%3B%0D%0Afunction%20test%3CT%3E(...values%3A%20T%5B%5D)%3A%20boolean%20%7B%0D%0A%20%20%20%20return%20values.length%20%3D%3D%3D%201%3B%0D%0A%7D%0D%0A%0D%0Adeclare%20function%20assertTrue(truth%3A%20true)%3A%20void%3B%0D%0A%0D%0Aconst%20noComma%20%3D%20test(%0D%0A%20%20%20%20'foo'%0D%0A)%0D%0AassertTrue(noComma)%3B%0D%0A%0D%0Aconst%20withComma%20%3D%20test(%0D%0A%20%20%20%20'foo'%2C%0D%0A)%3B%0D%0AassertTrue(withComma)%3B%0D%0A%0D%0AassertTrue(test(%0D%0A%20%20%20%20'foo'%2C%0D%0A))%3B%0D%0A)

**Related Issues:** #7279

**Comment by JoseLion** [comment]
Any news on this issue? The problem is affecting more than just the quick info tooltip. For example, if a lint tool uses the wrong inferred overload, it can cause an error/warning if that overload is marked as deprecated. I stumbled into precisely that scenario while using the `.map(..)` operator of [RxJS](https://github.com/ReactiveX/rxjs), which has the following signature (at the time):
```ts
export function map<T, R>(project: (value: T, index: number) => R): OperatorFunction<T, R>;
/** @deprecated Use a closure instead of a `thisArg`. Signatures accepting a `thisArg` will be removed in v8. */
export function map<T, R, A>(project: (this: A, value: T, index: number) => R, thisArg: A): OperatorFunction<T, R>;
```

Of course, the deprecated API will be removed eventually, but until that happens, Linters checking for deprecated code can give us a hard time:
```ts
const source = from([1, 2, 3, 4, 5]);

source.pipe(map(val => val + 10)); // complaint

source.pipe(
  map( // non-complaint: the deprecated overload is inferred
    val => val + 10,
  ),
);
```

I can try to help with a PR, but it'd be great if someone could point me in the right direction as this would be my first time contributing to TypeScript's codebase 🙂 

Thanks in advance!

**Comment by RyanCavanaugh** [comment]
Closing language service bugs related to the 6.0 implementation. For more information, see #62827
